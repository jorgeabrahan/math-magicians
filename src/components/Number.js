import PropTypes from 'prop-types';

export default function Number({ number, twoCols }) {
  return (
    <button type="button" className={twoCols ? 'c-span-2' : ''}>
      {number}
    </button>
  );
}

Number.propTypes = {
  number: PropTypes.number.isRequired,
  twoCols: PropTypes.bool,
};

Number.defaultProps = {
  twoCols: false,
};
