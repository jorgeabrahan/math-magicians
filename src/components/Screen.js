import PropTypes from 'prop-types';

export default function Screen({ content }) {
  return <p className="calculator__screen">{content}</p>;
}

Screen.propTypes = {
  content: PropTypes.string,
};

Screen.defaultProps = {
  content: '0',
};
