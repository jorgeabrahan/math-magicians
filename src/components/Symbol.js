import PropTypes from 'prop-types';

export default function Symbol({ symbol, orange, onClick }) {
  return (
    <button type="button" className={orange ? 'orange' : ''} onClick={() => onClick(symbol)}>
      {symbol}
    </button>
  );
}

Symbol.propTypes = {
  symbol: PropTypes.string.isRequired,
  orange: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
};

Symbol.defaultProps = {
  orange: false,
};
