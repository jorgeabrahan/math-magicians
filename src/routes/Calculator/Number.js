// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';

export default function Number({ number, twoCols, onClick }) {
  return (
    <button type="button" className={twoCols ? 'c-span-2' : ''} onClick={() => onClick(number)}>
      {number}
    </button>
  );
}

Number.propTypes = {
  number: PropTypes.number.isRequired,
  twoCols: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
};

Number.defaultProps = {
  twoCols: false,
};
