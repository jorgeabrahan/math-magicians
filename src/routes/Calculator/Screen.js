// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';

export default function Screen({ displayValue }) {
  const value = displayValue();
  return <p className="calculator__screen">{value === null ? '0' : value}</p>;
}

Screen.propTypes = {
  displayValue: PropTypes.func,
};

Screen.defaultProps = {
  displayValue: () => {},
};
