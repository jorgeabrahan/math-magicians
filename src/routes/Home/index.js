export default function Home() {
  return (
    <>
      <h2>Welcome to our page!</h2>
      <p>
        Lorem ipsum dolor sit amet consectetur, adipisicing elit. Error hic consectetur numquam unde
        minima quo odio qui, ab neque fuga! Iusto id ab provident praesentium? Perferendis
        exercitationem asperiores quaerat necessitatibus. Lorem ipsum dolor sit amet consectetur,
        adipisicing elit. Error hic consectetur numquam undeminima quo odio qui, ab neque fuga!
        Iusto id ab provident praesentium? Perferendis exercitationem asperiores quaerat
        necessitatibus.
      </p>
      <p>
        Lorem ipsum dolor sit amet consectetur, adipisicing elit. Error hic consectetur numquam unde
        minima quo odio qui, ab neque fuga! Iusto id ab provident praesentium? Perferendis
        exercitationem asperiores quaerat necessitatibus. Lorem ipsum dolor sit amet consectetur,
        adipisicing elit. Error hic consectetur numquam undeminima quo odio qui, ab neque fuga!
        Iusto id ab provident praesentium? Perferendis exercitationem asperiores quaerat
        necessitatibus.
      </p>
    </>
  );
}
