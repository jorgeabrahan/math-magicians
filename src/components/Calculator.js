import Number from './Number';
import Symbol from './Symbol';
import './calculator.css';
import Screen from './Screen';

export default function Calculator() {
  return (
    <main className="center">
      <div className="calculator">
        <Screen />
        <div className="row">
          <Symbol symbol="AC" />
          <Symbol symbol="+/-" />
          <Symbol symbol="%" />
          <Symbol symbol="/" orange />
        </div>

        <div className="row">
          <Number number={7} />
          <Number number={8} />
          <Number number={9} />
          <Symbol symbol="x" orange />
        </div>

        <div className="row">
          <Number number={4} />
          <Number number={5} />
          <Number number={6} />
          <Symbol symbol="-" orange />
        </div>

        <div className="row">
          <Number number={1} />
          <Number number={2} />
          <Number number={3} />
          <Symbol symbol="+" orange />
        </div>

        <div className="row">
          <Number number={0} twoCols />
          <Symbol symbol="." />
          <Symbol symbol="=" orange />
        </div>
      </div>
    </main>
  );
}
