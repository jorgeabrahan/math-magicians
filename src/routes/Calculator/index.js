import './styles/calculator.css';
import { useState } from 'react';
import Screen from './Screen';
import Number from './Number';
import Symbol from './Symbol';
import calculate from './logic/calculate';

export default function Calculator() {
  const [details, setDetails] = useState({
    total: null,
    next: null,
    operation: null,
  });
  const [isLastNumber, setIsLastNumber] = useState(false);
  const handleButtonClick = (buttonName) => {
    setDetails((current) => ({
      ...current,
      ...calculate(details, buttonName.toString().trim()),
    }));
    // Dots should be considered as numbers
    if (buttonName === '.') {
      setIsLastNumber(true);
      return;
    }
    setIsLastNumber(!!buttonName.toString().match(/[0-9]+/));
  };

  const displayValue = () => {
    // If the last button clicked was an operation
    if (!isLastNumber) return details.total;
    // If the last button clicked was '='
    if (details.total !== null && details.next === null && details.operation === null) {
      return details.total;
    }
    // In any other case
    return details.next;
  };
  return (
    <main className="center">
      <div className="calculator">
        <Screen displayValue={displayValue} />
        <div className="row">
          <Symbol symbol="AC" onClick={handleButtonClick} />
          <Symbol symbol="+/-" onClick={handleButtonClick} />
          <Symbol symbol="%" onClick={handleButtonClick} />
          <Symbol symbol="รท" onClick={handleButtonClick} orange />
        </div>

        <div className="row">
          <Number number={7} onClick={handleButtonClick} />
          <Number number={8} onClick={handleButtonClick} />
          <Number number={9} onClick={handleButtonClick} />
          <Symbol symbol="x" onClick={handleButtonClick} orange />
        </div>

        <div className="row">
          <Number number={4} onClick={handleButtonClick} />
          <Number number={5} onClick={handleButtonClick} />
          <Number number={6} onClick={handleButtonClick} />
          <Symbol symbol="-" onClick={handleButtonClick} orange />
        </div>

        <div className="row">
          <Number number={1} onClick={handleButtonClick} />
          <Number number={2} onClick={handleButtonClick} />
          <Number number={3} onClick={handleButtonClick} />
          <Symbol symbol="+" onClick={handleButtonClick} orange />
        </div>

        <div className="row">
          <Number number={0} onClick={handleButtonClick} twoCols />
          <Symbol symbol="." onClick={handleButtonClick} />
          <Symbol symbol="=" onClick={handleButtonClick} orange />
        </div>
      </div>
    </main>
  );
}
