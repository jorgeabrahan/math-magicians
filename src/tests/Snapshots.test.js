// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';
import Calculator from '../routes/Calculator/index';
import Screen from '../routes/Calculator/Screen';
import Number from '../routes/Calculator/Number';
import Symbol from '../routes/Calculator/Symbol';
import Home from '../routes/Home';
import Quote from '../routes/Quote';

describe('Snapshot tests for all components', () => {
  it('Should render the Calculator component correctly', () => {
    expect(renderer.create(<Calculator />).toJSON()).toMatchSnapshot();
  });
  it('Should render the Number component correctly', () => {
    expect(renderer.create(<Number />).toJSON()).toMatchSnapshot();
  });
  it('Should render the Symbol component correctly', () => {
    expect(renderer.create(<Symbol />).toJSON()).toMatchSnapshot();
  });
  it('Should render the Screen component correctly', () => {
    expect(renderer.create(<Screen />).toJSON()).toMatchSnapshot();
  });
  it('Should render the Home component correctly', () => {
    expect(renderer.create(<Home />).toJSON()).toMatchSnapshot();
  });
  it('Should render the Quote component correctly', () => {
    expect(renderer.create(<Quote />).toJSON()).toMatchSnapshot();
  });
});
