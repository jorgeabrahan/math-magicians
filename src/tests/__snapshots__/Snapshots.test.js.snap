// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests for all components Should render the Calculator component correctly 1`] = `
<section
  className="calculator-container"
>
  <h2
    className="subtitle"
  >
    Let's do some math!
  </h2>
  <div
    className="calculator"
  >
    <p
      className="calculator__screen"
    >
      0
    </p>
    <div
      className="row"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="orange"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="orange"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="orange"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="orange"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="c-span-2"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="orange"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</section>
`;

exports[`Snapshot tests for all components Should render the Home component correctly 1`] = `
Array [
  <h2
    className="subtitle"
  >
    Welcome to our page!
  </h2>,
  <p
    style={
      Object {
        "marginBottom": "1.5rem",
      }
    }
  >
    Lorem ipsum dolor sit amet consectetur, adipisicing elit. Error hic consectetur numquam unde minima quo odio qui, ab neque fuga! Iusto id ab provident praesentium? Perferendis exercitationem asperiores quaerat necessitatibus. Lorem ipsum dolor sit amet consectetur, adipisicing elit. Error hic consectetur numquam undeminima quo odio qui, ab neque fuga! Iusto id ab provident praesentium? Perferendis exercitationem asperiores quaerat necessitatibus.
  </p>,
  <p>
    Lorem ipsum dolor sit amet consectetur, adipisicing elit. Error hic consectetur numquam unde minima quo odio qui, ab neque fuga! Iusto id ab provident praesentium? Perferendis exercitationem asperiores quaerat necessitatibus. Lorem ipsum dolor sit amet consectetur, adipisicing elit. Error hic consectetur numquam undeminima quo odio qui, ab neque fuga! Iusto id ab provident praesentium? Perferendis exercitationem asperiores quaerat necessitatibus.
  </p>,
]
`;

exports[`Snapshot tests for all components Should render the Number component correctly 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
/>
`;

exports[`Snapshot tests for all components Should render the Quote component correctly 1`] = `
<p
  className="quote"
>
  Mathematics is not about numbers, equations, computations, or algorithms; it is about understanding. ~William Paul Thurston
</p>
`;

exports[`Snapshot tests for all components Should render the Screen component correctly 1`] = `
<p
  className="calculator__screen"
/>
`;

exports[`Snapshot tests for all components Should render the Symbol component correctly 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
/>
`;
